lotman/lotman_linux.go

```diff
@@ -314,9 +333,580 @@ func getFederationIssuer() (string, error) {
+		if totalDedicatedGB > bytesToGigabytes(hwm) {
+			return errors.New(fmt.Sprintf("the sum of all lots' dedicatedGB values exceeds the high watermark of %s. This would allow the cache to purge namespaces using less than their dedicated quota", hwmStr))
+		}
+	}
+
+	return nil
+}
+
+// HWM and LWM values may be a percentage (e.g. 95) indicating the amount of available disk
+// to treat as the watermark, or they may be a suffixed byte value (e.g. 100G). We need this
+// information in bytes to calculate the amount of space to allocate to each lot.
+func convertWatermarkToBytes(value string, totalDiskSpace uint64) (uint64, error) {
+    suffixMultipliers := map[string]uint64{
+        "k": 1000,
+        "m": 1000 * 1000,
+        "g": 1000 * 1000 * 1000,
+        "t": 1000 * 1000 * 1000 * 1000,
+    }
+
+    // Check if the value has a suffix
+    if len(value) > 1 {
+        suffix := strings.ToLower(string(value[len(value)-1]))
+        if multiplier, exists := suffixMultipliers[suffix]; exists {
+            number, err := strconv.ParseUint(value[:len(value)-1], 10, 64)
+            if err != nil {
+                return 0, err
+            }
+            return number * multiplier, nil
+        }
+    }
+
+    // If no suffix, treat as percentage
+    percentage, err := strconv.ParseFloat(strings.TrimSuffix(value, "%"), 64)
+    if err != nil {
+        return 0, err
+    }
+    return uint64((percentage / 100) * float64(totalDiskSpace)), nil
+}
+
+
+// Divide the remaining space among lots' dedicatedGB values -- we don't ever want to
+// dedicate more space than we have available, as indicated by the HWM of the cache. This is because
+// our model is that as long as a lot stays under its dedicated GB, its data is safe in the cache --
+// If the sum of each lot's dedicated GB exceeds the HWM, the cache may purge data without a single lot
+// exceeding it's quota. BAD!
+//
+// Opportunistic space can (and should) be overallocated, so unless explicitly set, each lot will have
+// dedicatedGB + opportunisticGB = HWM. This isn't maxed out to the total disk space, because otherwise
+// no lot could ever exceed its opportunistic storage and we'd lose some of the capabilities to reason about
+// how greedy the lot is.
+func divideRemainingSpace(lotMap *map[string]Lot, totalDiskSpaceB uint64) error {
+	hwmStr := param.Cache_HighWaterMark.GetString()
+	if hwmStr == "" {
+		return errors.New("high watermark is not set in the cache configuration")
+	}
+	hwm, err := convertWatermarkToBytes(hwmStr, totalDiskSpaceB)
+	if err != nil {
+		return errors.Wrap(err, "error converting high watermark to byte value for Lotman")
+	}
+	remainingToHwmB := hwm
+
+    // first iterate through all lots and subtract from our total space any amount
+    // that's already been allocated. Note which lots have an unset value, as we'll
+    // need to return to them.
+    returnToKeys := make([]string, 0, len(*lotMap))
+	for key, lot := range *lotMap {
```

Comment:
During this loop, isn't it possible that remainingToHwmB could become negative? What occurs there? You say you don't ever want to dedicate more space than we have available, but I'm not seeing any check that is occurring here. 

EDIT: Ah, I see the check is later in Validation function.  Either add a comment saying this, as the above is misleading (you specify it's bad but don't clarify it's checked elsewhere) OR take this opportunity to see that the values conflict here (if remainingToHwmB ever gets < 0) and return the appropriate error and exit early.

Reply:
I agree it's a good idea to catch it early. I'll add a check here in addition to referencing the validation that happens later.